### ISSUE TITLE
Using logstash queries to enrich log entries

### ISSUE THREAD

**Issue by fastlorenzo** [outflanknl/RedELK#36]
Instead of using the ruby script to add some metadata (like the beacon details) to log entries, it's possible to use logstash's elasticsearch filter and directly query the elasticsearch database.

I'm currently doing this for several points:

1. Adding beacon data from beacon_id:

After [elkserver/logstash/conf.d/50-c2-cobaltstrike.conf#L89](https://github.com/outflanknl/RedELK/blob/2233941fc9bb1430d44973e70fec39c513b74aa6/elkserver/logstash/conf.d/50-c2-cobaltstrike.conf#L89) and [elkserver/logstash/conf.d/50-c2-cobaltstrike.conf#L253](https://github.com/outflanknl/RedELK/blob/2233941fc9bb1430d44973e70fec39c513b74aa6/elkserver/logstash/conf.d/50-c2-cobaltstrike.conf#L253) added the following:

```
    # enrich the beacon data using the retrieved beacon_id
    elasticsearch {
      hosts => "127.0.0.1"
      index => "beacondb"
      query => "beacon_id:%{beacon_id}"
      fields => {
        "target_hostname" => "target_hostname"
        "target_ipext" => "target_ipext"
        "target_os" => "target_os"
        "target_osversion" => "target_osversion"
        "target_osbuild" => "target_osbuild"
        "target_pid" => "target_pid"
        "target_user" => "target_user"
      }
      result_size => 1
      enable_sort => false
      user => ["logstash"]
      password => ["{{elk_logstash_system_password}}"]
    }
```

This does the lookup directly from ES and can then be removed from the ruby script used to enrich the data.

2. Added new indices for the different IP lists:

I've added new indices to have the different IP lists stored, and after [elkserver/logstash/conf.d/20-redir-haproxy.conf#L154](https://github.com/outflanknl/RedELK/blob/2233941fc9bb1430d44973e70fec39c513b74aa6/elkserver/logstash/conf.d/20-redir-haproxy.conf#L154), I've added the following:

 ```
      # matches the source ip with any ip in the different ip lists
      elasticsearch {
        hosts => "127.0.0.1"
        index => "iplist_*"
        query => "ip:%{src_ip}"
        fields => { "type" => "iplist" }
        result_size => 1
        enable_sort => false
        user => ["logstash"]
        password => ["{{elk_logstash_system_password}}"]
      }

      # unknown IP
      if ![iplist] {
        # checking with greynoise
        ruby { 
          path => "/usr/share/redelk/bin/greynoise.rb"
        }
      } else {
        mutate {
          add_tag => [ "iplist_%{iplist}" ]
        }
      }
```

This allows normally to get rid of ´elkserver/scripts/enrich.py´ and do most of the processing in logstash. Please note that you still need to maintain your different ´iplist_*´ indices (like tor, redteam ips, etc.).

I'll be happy to integrate those changes in a PR if you see a use of this.

I was actually wondering if there is a specific reason of not using logstash to do that processing?

**Comment by MarcOverIP** [comment]
I guess you mean the python (enrich.py) script instead of 'ruby script'? Ruby is only used for inserting links to URLs visible  in Kibana.

The approach you mention was our first thought as well and we have gone down this route in our very first versions of RedELK. However we experienced a series of issues, mainly:
1. if the first "metadata" line of CS is not yet available in ES, the lookup will fail. In our testing this happened too much. Afaik this wasnt a resource issue, but an issue when the filebeat-logstash link was down and they were playing catchup.
2. There is no way of historically modifying things. If the lookup failed, the grok parsing failed and will keep as failed forever. This really comes to play in the 2nd point of your post, the IP lists. At least in our operations we have increasing knowledge of the different sets of IP addresses. We want those historically changed as well. IP is just an example. We can use enrich.py on other info as well that we like to add during the ops (including other C2 frameworks). It is a great template for historically modifying data.

Now, it has been a time since we looked at his (ES version 5.x), and perhaps things have changed. A solution for the first issue could as well be to start using a setup with Kafka and KSQL. But tbh it does seem a bit as overkill and complex.
Yet at the same time I can imagine that the current enrich.py - although it works - might seem overkill and overly complex as well. It might not give the same level of flexibility we really enjoy if you aren't comfortable with the code...?

Happy with your thoughts and overall happy with you thinking along with the project! Im open for discussion, just wanted to let you know we have looked at this some time ago.
